plugins/soundsourcem4a/audiosourcem4a.cpp

```diff
@@ -0,0 +1,307 @@
+            return ERR;
+        }
+        faacDecConfigurationPtr pDecoderConfig = faacDecGetCurrentConfiguration(
+                m_hDecoder);
+        pDecoderConfig->outputFormat = FAAD_FMT_FLOAT; /* 32-bit float */
+        pDecoderConfig->downMatrix = 1; /* 5.1 -> stereo */
+        pDecoderConfig->defObjectType = LC;
+        if (!faacDecSetConfiguration(m_hDecoder, pDecoderConfig)) {
+            qWarning() << "Failed to configure AAC decoder!";
+            return ERR;
+        }
+    }
+
+    u_int8_t* configBuffer = NULL;
+    u_int32_t configBufferSize = 0;
+    if (!MP4GetTrackESConfiguration(m_hFile, m_trackId, &configBuffer,
+            &configBufferSize)) {
+        /* failed to get mpeg-4 audio config... this is ok.
+         * faacDecInit2() will simply use default values instead.
+         */
+        qWarning()
+                << "Failed to read the MP4 audio configuration. Continuing with default values.";
+    }
+
+    SAMPLERATE_TYPE sampleRate;
+    unsigned char channelCount;
+    if (0 > faacDecInit2(m_hDecoder, configBuffer, configBufferSize,
+                    &sampleRate, &channelCount)) {
+        free(configBuffer);
+        qWarning() << "Failed to initialize the AAC decoder!";
+        return ERR;
+    } else {
+        free(configBuffer);
+    }
+
+    setChannelCount(channelCount);
+    setFrameRate(sampleRate);
+    setFrameCount(m_maxSampleId * kFramesPerSampleBlock);
+
+    const SampleBuffer::size_type prefetchSampleBufferSize = (kSampleIdPrefetchCount + 1) * frames2samples(kFramesPerSampleBlock);
+    SampleBuffer(prefetchSampleBufferSize).swap(m_prefetchSampleBuffer);
+
+    // invalidate current frame index
+    m_curFrameIndex = getFrameCount();
+    // seek to beginning of file
+    if (0 != seekFrame(0)) {
+        qWarning() << "Failed to seek to the beginning of the file!";
+        return ERR;
+    }
+
+    return OK;
+}
+
+void AudioSourceM4A::close() throw() {
+    if (m_hDecoder) {
+        NeAACDecClose(m_hDecoder);
+        m_hDecoder = NULL;
+    }
+    if (MP4_INVALID_FILE_HANDLE != m_hFile) {
+        MP4Close(m_hFile);
+        m_hFile = MP4_INVALID_FILE_HANDLE;
+    }
+    m_trackId = MP4_INVALID_TRACK_ID;
+    m_maxSampleId = MP4_INVALID_SAMPLE_ID;
+    m_curSampleId = MP4_INVALID_SAMPLE_ID;
+    m_inputBuffer.clear();
+    m_inputBufferOffset = 0;
+    m_inputBufferLength = 0;
+    m_curFrameIndex = 0;
+    Super::reset();
+}
+
+bool AudioSourceM4A::isValidSampleId(MP4SampleId sampleId) const {
+    return (sampleId >= kMinSampleId) && (sampleId <= m_maxSampleId);
+}
+
+AudioSource::diff_type AudioSourceM4A::seekFrame(diff_type frameIndex) {
+    if (m_curFrameIndex != frameIndex) {
+        const MP4SampleId sampleId = kMinSampleId
+                + (frameIndex / kFramesPerSampleBlock);
+        if (!isValidSampleId(sampleId)) {
+            return m_curFrameIndex;
+        }
+        if ((m_curSampleId != sampleId) || (frameIndex < m_curFrameIndex)) {
```

Comment:
I adopted the following definition for the term "frame" in Mixxx: http://techpubs.sgi.com/library/tpl/cgi-bin/getdoc.cgi?coll=0650&db=bks&srch=&fname=/SGI_Developer/DMSDK_PG/sgi_html/ch08.html

"A sample frame is a set of audio samples that are coincident in time. A sample frame for mono data is a single sample".

May it help if we change the term from "frame" to "sample frame"? I decided to drop the "sample" prefix to have a short and catchy term.

Each format and even each decoding library have its own understanding of "sample" and "frame", very confusing. M4A/FAAD for example uses the term "sample" for a whole block of sample frames that is addressed by MP4SampleId. At least all those naming conflicts will be isolated in the implementations of the AudioSource sub-classes.